File: frontend/investigations.js
/* Isolated investigations never write to fleet history. */
(() => {
  'use strict';
  let report = null;
  let charts = [];
  const escape = value => String(value ?? '').replace(/[&<>"']/g, char => ({'&':'&amp;','<':'&lt;','>':'&gt;','"':'&quot;',"'":'&#39;'}[char]));
  const number = (value, unit = '') => value == null ? 'Not available' : typeof value === 'boolean' ? (value ? 'Yes' : 'No') : typeof value !== 'number' ? String(value) : `${value.toLocaleString(undefined, {maximumFractionDigits: 2})}${unit}`;
  const when = (value, basis) => value == null ? 'Not available' : basis === 'utc' ? new Date(value * 1000).toISOString() : `${number(value)} s uptime`;

  function resultHtml(data, plots = false) {
    return `<h2>${escape(data.status)}</h2><p class="muted">This investigation is separate from fleet history and learning.</p>` + data.results.map((result, index) => {
      const a = result.analysis, m = a.metrics, h = a.health || {};
      const recommendations = Array.isArray(a.findings?.recommendations) ? a.findings.recommendations : [];
      const nextStep = typeof recommendations[0] === 'string' ? recommendations[0] : recommendations[0]?.action || recommendations[0]?.title;
      const reasons = h.reasons || [];
      return `<section class="investigation-result"><h3>${escape(result.name)}</h3>
        <p>${escape(result.files.join(', '))}</p>
        <div class="investigation-metrics"><div><small>GC health</small><strong>${escape(h.grade || '?')} / ${escape(h.status || 'unknown')}</strong></div>
        <div><small>Events</small><strong>${number(m.event_count)}</strong></div>
        <div><small>Longest pause</small><strong>${number(m.max_pause_ms, ' ms')}</strong></div>
        <div><small>Time in GC</small><strong>${number(m.pct_time_in_gc, '%')}</strong></div></div>
        <p><strong>Time basis:</strong> ${result.time_basis === 'utc' ? 'UTC' : 'JVM uptime only'}<br>${escape(when(result.quality.start, result.time_basis))} to ${escape(when(result.quality.end, result.time_basis))}</p>
        <h4>What we found</h4><p>${escape(reasons[0] || (m.event_count ? 'No scored GC warning in the supplied sample.' : 'No supported GC events were found.'))}</p>
        <h4>Why it matters</h4><p>${m.event_count ? 'GC pauses can delay application work. Kafka service impact is not confirmed by this log alone.' : 'Health cannot be assessed from this file.'}</p>
        <h4>Next step</h4><p>${escape(nextStep || 'Compare a representative busy period with host and Kafka metrics.')}</p>
        ${(a.warnings || []).map(w => `<p class="investigation-warning">${escape(w)}</p>`).join('')}
        ${plots && a.timeline?.length ? `<div class="investigation-chart"><canvas id="investigation-plot-${index}" aria-label="Observed GC pause timeline" role="img"></canvas></div>` : ''}
        <details><summary>All observed metrics</summary><table class="investigation-table"><tbody>${Object.entries(m).map(([k,v])=>`<tr><th>${escape(k.replaceAll('_',' '))}</th><td>${escape(number(v))}</td></tr>`).join('')}</tbody></table></details>
      </section>`;
    }).join('') + `<section class="investigation-result"><h3>GC and host evidence</h3><p>${escape(data.correlation.summary)}</p>
      <p>Matched minutes: ${number(data.correlation.matched_minutes)}. CPU / pause association: ${number(data.correlation.cpu_pause_correlation)}.</p>
      ${data.correlation.sampling_note ? `<p class="muted">${escape(data.correlation.sampling_note)}</p>` : ''}
      ${data.host ? `<p>Host samples: ${number(data.host.metrics.sample_count)}. CPU busy: ${number(data.host.metrics.cpu_busy_pct_avg, '%')}. Available memory: ${number(data.host.metrics.mem_avail_mb_avg, ' MB')}.</p>${(data.host.warnings || []).map(w=>`<p class="investigation-warning">${escape(w)}</p>`).join('')}` : ''}</section>`;
  }

  async function filePayload(file) {
    if (file.size > 4 * 1024 * 1024) throw new Error('Each upload must be 4 MiB or smaller.');
    if (/\.gz$/i.test(file.name)) {
      const bytes = new Uint8Array(await file.arrayBuffer());
      let binary = '';
      for (let offset = 0; offset < bytes.length; offset += 8192) binary += String.fromCharCode(...bytes.subarray(offset, offset + 8192));
      return {name: file.name, content: btoa(binary), encoding: 'gzip-base64'};
    }
    return {name: file.name, content: await file.text(), encoding: 'text'};
  }

  function destroyCharts() { charts.forEach(chart => chart.destroy()); charts = []; }

  function drawCharts(data) {
    if (!window.Chart) return;
    data.results.forEach((result, index) => {
      const canvas = document.getElementById(`investigation-plot-${index}`);
      if (!canvas) return;
      charts.push(new Chart(canvas, {type: 'line', data: {datasets: [{label: 'GC pause (ms)',
        data: result.analysis.timeline.map(p=>({x:p.t, y:p.pause_ms})), borderColor:'#0a85c2', pointRadius:1, borderWidth:1.5}]},
        options: {responsive:true, maintainAspectRatio:false, animation:false, parsing:false,
          scales: {x:{type:'linear', ticks:{maxTicksLimit:6, callback:value=>result.time_basis==='utc' ? new Date(value*1000).toISOString().slice(11,19) : `${value}s`}, title:{display:true,text:result.time_basis==='utc'?'UTC':'JVM uptime (seconds)'}}, y:{beginAtZero:true}},
          plugins:{legend:{display:false}, tooltip:{callbacks:{title:items=>when(items[0].parsed.x,result.time_basis)}}}}}));
    });
  }

  window.openInvestigation = function() {
    let dialog = document.getElementById('investigation-dialog');
    if (!dialog) {
      const stylesheet = document.createElement('link');
      stylesheet.rel = 'stylesheet'; stylesheet.href = '/assets/investigations.css'; document.head.append(stylesheet);
      dialog = document.createElement('dialog'); dialog.id = 'investigation-dialog';
      dialog.innerHTML = `<header><div><h2>Analyze logs</h2><span class="muted">One-time investigation</span></div><button type="button" id="investigation-close" aria-label="Close investigation" title="Close">&times;</button></header>
        <form id="investigation-form"><div class="investigation-fields">
        <label>GC logs<input id="investigation-files" type="file" multiple required accept=".log,.txt,.gz,.0,.1,.2,.3,.4,.5,.6,.7,.8,.9"></label>
        <label>Host SAR export (optional)<input id="investigation-sar" type="file" accept=".json,.txt,.gz"></label>
        <label class="investigation-check"><input id="investigation-group" type="checkbox"> Files are rotations from the same JVM</label>
        <label>JVM start time (optional)<input id="investigation-anchor" type="text" placeholder="2026-09-12T00:00:00Z" title="ISO-8601 timestamp including timezone"></label>
        <label>SAR date (optional)<input id="investigation-date" type="date"></label>
        <label>SAR report timezone<input id="investigation-zone" type="text" value="UTC" title="Timezone used when the SAR report was exported"></label></div>
        <div class="investigation-actions"><button type="submit" id="investigation-submit">Analyze</button><button type="button" id="investigation-export" disabled>Download report</button><span id="investigation-status" role="status" aria-live="polite"></span></div></form>
        <div id="investigation-output"></div>`;
      document.body.append(dialog);
      document.getElementById('investigation-close').onclick = () => dialog.close();
      document.getElementById('investigation-export').onclick = () => {
        if (!report) return;
        const html = `<!doctype html><html lang="en"><meta charset="utf-8"><meta name="viewport" content="width=device-width"><title>GC investigation</title><style>body{font:15px system-ui;color:#0a2540;max-width:1000px;margin:32px auto;padding:0 24px}section{border-top:1px solid #d8e1ea;padding:20px 0}table{border-collapse:collapse;width:100%}th,td{text-align:left;padding:6px;border-bottom:1px solid #d8e1ea}small{display:block}strong{margin-right:24px}.investigation-metrics{display:flex;gap:24px;flex-wrap:wrap}p,th,td{overflow-wrap:anywhere}</style><h1>GC investigation</h1><p>${escape(report.generated_at)}</p>${resultHtml(report)}<p>No uploaded data was added to fleet history or learning baselines.</p></html>`;
        const url = URL.createObjectURL(new Blob([html], {type:'text/html'}));
        const link = document.createElement('a'); link.href=url; link.download='gc-investigation.html'; document.body.append(link); link.click(); link.remove(); setTimeout(()=>URL.revokeObjectURL(url),10000);
      };
      document.getElementById('investigation-form').onsubmit = async event => {
        event.preventDefault();
        const button = document.getElementById('investigation-submit'), status = document.getElementById('investigation-status');
        const output = document.getElementById('investigation-output'), download = document.getElementById('investigation-export');
        button.disabled=true; download.disabled=true; status.textContent='Analyzing...'; report=null; destroyCharts(); output.replaceChildren();
        try {
          const selected = Array.from(document.getElementById('investigation-files').files);
          if (!selected.length || selected.length>12) throw new Error('Select between 1 and 12 GC files.');
          const files=[];
          for (const file of selected) files.push(await filePayload(file));
          if (document.getElementById('investigation-group').checked) files.forEach(file=>file.group='uploaded-jvm');
          const body={files};
          const sar=document.getElementById('investigation-sar').files[0];
          if (sar) body.sar=await filePayload(sar);
          const anchor=document.getElementById('investigation-anchor').value.trim();
          if(anchor) body.start_time=anchor;
          const date=document.getElementById('investigation-date').value;
          if(date) body.report_date=date;
          body.report_timezone=document.getElementById('investigation-zone').value.trim();
          const response=await fetch('/api/investigations/analyze',{method:'POST',headers:{'Content-Type':'application/json'},body:JSON.stringify(body)});
          const data=await response.json();
          if(!response.ok) throw new Error(data.detail || 'Analysis could not complete.');
          report=data; output.innerHTML=resultHtml(data,true); drawCharts(data); download.disabled=false; status.textContent='Analysis complete';
        } catch(error) {status.textContent=error.message;}
        finally {button.disabled=false;}
      };
    }
    dialog.showModal();
  };
  if (typeof module === 'object' && module.exports) module.exports = {resultHtml};
})();

File: gcanalyzer/investigations.py
"""Bounded, stateless analysis of operator-supplied logs. No fleet-store access."""

from __future__ import annotations

import base64
import binascii
import gzip
import io
import statistics
from datetime import datetime, timezone
from zoneinfo import ZoneInfo, ZoneInfoNotFoundError

from . import analyzer, parser, sar_analyzer, sar_parser

MAX_REQUEST_BYTES = 12 * 1024 * 1024
MAX_FILE_BYTES = 4 * 1024 * 1024
MAX_TOTAL_BYTES = 8 * 1024 * 1024
MAX_FILES = 12
BUCKET_SECONDS = 60


def _decode_file(file: dict) -> tuple[str, str, int]:
    if not isinstance(file, dict):
        raise ValueError("Each file must have a name and content.")
    name, content = file.get("name", "upload"), file.get("content")
    if not isinstance(name, str) or not name or len(name) > 255 or not isinstance(content, str):
        raise ValueError("Provide a short file name and text content.")
    encoding = file.get("encoding", "text")
    if encoding == "text":
        raw = content.encode("utf-8")
    elif encoding == "gzip-base64":
        try:
            compressed = base64.b64decode(content, validate=True)
            with gzip.GzipFile(fileobj=io.BytesIO(compressed)) as archive:
                raw = archive.read(MAX_FILE_BYTES + 1)
        except (binascii.Error, OSError, EOFError) as exc:
            raise ValueError(f"Invalid gzip file: {name}") from exc
    else:
        raise ValueError("Supported encodings are text and gzip-base64.")
    if len(raw) > MAX_FILE_BYTES:
        raise ValueError(f"File exceeds the {MAX_FILE_BYTES // 1024 // 1024} MiB expanded size limit.")
    try:
        text = raw.decode("utf-8-sig")
    except UnicodeDecodeError as exc:
        raise ValueError(f"{name} is not UTF-8 text. Export binary SAR data as sadf JSON first.") from exc
    if "\x00" in text:
        raise ValueError(f"{name} contains binary data. Supply a text or JSON export.")
    return name, text, len(raw)


def _start_epoch(value: str | None) -> float | None:
    if value is None or value == "":
        return None
    try:
        dt = datetime.fromisoformat(value.replace("Z", "+00:00"))
        if dt.tzinfo is None or dt.utcoffset() is None:
            raise ValueError
        return dt.timestamp()
    except (ValueError, TypeError, AttributeError, OverflowError) as exc:
        raise ValueError("JVM start time must be ISO-8601 with a timezone, for example 2026-09-12T00:00:00Z.") from exc


def _correlation(parsed, host) -> dict:
    unavailable = {"status": "unavailable", "matched_minutes": 0, "series": []}
    if host is None or not host.samples:
        return {**unavailable, "summary": "Host data not supplied or not recognized."}
    if not parsed.events or any(e.timestamp is None for e in parsed.events):
        return {**unavailable, "summary": "A confirmed UTC timeline is needed to compare GC and host data."}
    gc, sar = {}, {}
    for event in parsed.events:
        if event.is_stw:
            gc.setdefault(int(event.timestamp // BUCKET_SECONDS), []).append(event.pause_ms)
    for sample in host.samples:
        sar.setdefault(int(sample.ts // BUCKET_SECONDS), []).append(sample)
    series = []
    for minute in sorted(gc.keys() & sar.keys()):
        samples = sar[minute]
        cpu = [100 - s.cpu_idle_pct for s in samples if s.cpu_idle_pct is not None]
        io_wait = [s.cpu_iowait_pct for s in samples if s.cpu_iowait_pct is not None]
        memory = [s.mem_avail_mb for s in samples if s.mem_avail_mb is not None]
        series.append({"t": minute * BUCKET_SECONDS, "pause_max_ms": max(gc[minute]),
                       "cpu_busy_pct": statistics.fmean(cpu) if cpu else None,
                       "iowait_pct": statistics.fmean(io_wait) if io_wait else None,
                       "mem_available_mb": statistics.fmean(memory) if memory else None})
    usable = [r for r in series if r["cpu_busy_pct"] is not None]
    coefficient = None
    if len(usable) >= 6:
        try:
            coefficient = round(statistics.correlation([r["pause_max_ms"] for r in usable],
                                                       [r["cpu_busy_pct"] for r in usable]), 3)
        except statistics.StatisticsError:
            pass
    return {"status": "available" if series else "unavailable", "matched_minutes": len(series),
            "series": series[:1500], "cpu_pause_correlation": coefficient,
            "summary": ("GC and host samples overlap. Association does not confirm a cause or Kafka impact."
                        if series else "No overlapping GC and host samples in the same UTC minute."),
            "sampling_note": "Only minutes containing both observed GC events and host samples are compared; gaps are not zero-filled."}


def analyze_upload(data: dict) -> dict:
    if not isinstance(data, dict):
        raise ValueError("Provide a JSON object containing files.")
    files = data.get("files")
    if not isinstance(files, list) or not 1 <= len(files) <= MAX_FILES:
        raise ValueError(f"Provide between 1 and {MAX_FILES} GC files.")
    anchor = _start_epoch(data.get("start_time"))
    groups: dict[str, list] = {}
    total = 0
    for index, file in enumerate(files):
        name, text, size = _decode_file(file)
        total += size
        if total > MAX_TOTAL_BYTES:
            raise ValueError("Combined expanded upload exceeds the 8 MiB limit.")
        group = file.get("group")
        if group is not None and (not isinstance(group, str) or not group.strip() or len(group) > 255):
            raise ValueError("A JVM group must be a short non-empty name.")
        # Files remain independent unless the operator explicitly identifies a shared JVM.
        key = f"group:{group}" if group else f"file:{index}"
        groups.setdefault(key, []).append((name, text))
    host = None
    if data.get("sar") is not None:
        if len(groups) != 1:
            raise ValueError("Host data can only be compared with one JVM group at a time.")
        name, text, size = _decode_file(data["sar"])
        total += size
        if total > MAX_TOTAL_BYTES:
            raise ValueError("Combined expanded upload exceeds the 8 MiB limit.")
        report_date = data.get("report_date")
        if report_date is not None:
            try:
                datetime.strptime(report_date, "%Y-%m-%d")
            except (ValueError, TypeError) as exc:
                raise ValueError("SAR report date must be YYYY-MM-DD.") from exc
        report_timezone = data.get("report_timezone", "UTC")
        try:
            if not isinstance(report_timezone, str) or len(report_timezone) > 100:
                raise ValueError
            if report_timezone != "UTC":
                ZoneInfo(report_timezone)
        except (ValueError, ZoneInfoNotFoundError) as exc:
            raise ValueError("SAR timezone must be UTC or a valid IANA timezone name.") from exc
        host = sar_parser.parse(text, node_id=name, report_date=report_date, report_timezone=report_timezone)
    if total > MAX_TOTAL_BYTES:
        raise ValueError("Combined expanded upload exceeds the 8 MiB limit.")
    results, logs = [], []
    for group_files in groups.values():
        name = group_files[0][0]
        text = "\n".join(content for _, content in group_files)
        parsed = parser.parse(text, node_id=name, **({"start_time": anchor} if anchor is not None else {}))
        result = analyzer.analyze(parsed)
        if not result["metrics"]["stw_count"]:
            for key in ("avg_pause_ms", "max_pause_ms", "p50_pause_ms", "p95_pause_ms", "p99_pause_ms"):
                result["metrics"][key] = None
        absolute = bool(parsed.events) and all(e.timestamp is not None for e in parsed.events)
        times = [e.timestamp if absolute else e.uptime for e in parsed.events]
        times = [t for t in times if t is not None]
        quality = {"state": "missing" if not parsed.events else ("available" if absolute and result["health"]["score"] is not None else "partial"),
                   "event_count": len(parsed.events), "start": min(times) if times else None,
                   "end": max(times) if times else None, "warnings": list(parsed.warnings)}
        results.append({"name": name, "files": [name for name, _ in group_files],
                        "time_basis": "utc" if absolute else "relative", "quality": quality, "analysis": result})
        logs.append(parsed)
    correlation = _correlation(logs[0], host) if len(logs) == 1 else {
        "status": "unavailable", "matched_minutes": 0, "series": [], "summary": "Select one JVM group for host comparison."}
    return {"status": "Analysis complete", "persisted": False, "generated_at": datetime.now(timezone.utc).isoformat(),
            "results": results, "host": sar_analyzer.analyze(host) if host else None, "correlation": correlation,
            "limitations": ["Uploaded data is not added to fleet history or learning baselines.",
                            "Kafka service impact cannot be confirmed without Kafka metrics."]}
